gtda/utils/validation.py

```diff
@@ -76,143 +51,103 @@ def check_graph(X):
+            else:  # List, tuple or ndarray type
+                for i, parameter_elem in enumerate(parameter):
+                    _validate_params_single(
+                        parameter_elem, ref_of, f"{name}[{i}]")
+
+
+def validate_params(parameters, references, exclude=None):
+    """Function to automate the validation of (hyper)parameters.
 
     Parameters
     ----------
     parameters : dict, required
-        Dictionary in which the keys are hyperparameter names and the
-        corresponding values are hyperparameter values.
+        Dictionary in which the keys parameter names (as strings) and the
+        corresponding values are parameter values. Unless `exclude` (see
+        below) contains some of the keys in this dictionary, all parameters
+        are checked against `references`.
 
     references : dict, required
-        Dictionary in which the keys are hyperparameter names and the
-        corresponding values are lists. The first element of that list is a
-        type. If that type is not an iterable, the second element can be one
-        of:
-
-        - ``None``, when only the type should be checked.
-        - A tuple of two numbers, when the type is numerical. In this case,
-          the first (resp. second) entry in the tuple defines a lower
-          (resp. upper) bound constraining the value of the corresponding
-          hyperparameter.
-        - A list containing all possible allowed values for the
-          corresponding hyperparameter.
-
-        If that type is an iterable, the second element is a list that provides
-        information to validate each element of that iterable. The first
-        element of that list is the type of the elements of the iterable and
-        the second element of that list can be one of:
-
-        - ``None``, when only the type of the iterable elements should be
-          checked.
-        - A tuple of two numbers, when the type is numerical. In this case,
-          the first (resp. second) entry in the tuple defines a lower
-          (resp. upper) bound constraining the value of the corresponding
-          iterable element.
-        - A list containing all possible allowed values for the
-          corresponding iterable element.
+        Dictionary in which the keys are parameter names (as strings). The
+        value ``reference`` corresponding to a key ``name`` is a dictionary
+        responsible for checking the value ``parameter`` in `parameters`
+        corresponding to ``name``, and containing any of the following keys:
+
+        - ``'type'``, mapping to a class or tuple of classes. ``parameter``
+          is checked to be an instance of this class or tuple of classes.
+
+        - ``'in'``, when the value of ``'type'`` is not one of ``list``,
+          ``tuple``, ``numpy.ndarray`` or ``dict``. If the corresponding value
+          is ``ref_in`` the following check is performed on ``parameter``:
+          ``parameter in ref_in``.
```

Comment:
I don't understand the `ref_in` part.

Reply:
Bad language. I meant to say something like "let's call the corresponding value ``ref_in``, then...". I'll reword for clarity.

Reply:
Done.

Reply:
In general, `ref_in` is a variable that you define inside the function and is not visible. What about:
```
Let us call the constraints corresponding to the current parameter `ref`, that is
`ref = references['name']` and `param=parameters['name']`.
(...)
- ``'in'``: The following check is performed:
                ``param in ref['in']``. Used only when `ref['type']` is not one
               of ``list``, ``tuple``, ``numpy.ndarray`` or ``dict``.

- ``'of'``: Used when `ref['type']` is not one of ``list``, ``tuple``,
                ``numpy.ndarray`` or ``dict``.Then, `param` is an iterable,
                whose elements are check against the dictionary `ref['of']`.
```

Reply:
So if I understand correctly, the point is that you find expressions such as "The value ``reference`` corresponding to a key ``name``" and "`'in'``, mapping to a dictionary ``ref_in``" to be not sufficiently explicit, and would prefer to change them to e.g. "Let ``name`` denote a key and ``reference`` the corresponding value" -- more explicit. Correct?

Reply:
Yes, indeed

Reply:
Ok, I'll make changes in this spirit.

Reply:
Done, let me know if the new version suits.